src/persist-txn/src/txn_read.rs

```diff
@@ -343,16 +428,129 @@ where
+    C: TxnsCodec,
+{
+    /// Creates a [TxnsSubscribeTask] reading from the given txn shard that
+    /// forwards updates (entries and progress) to the given `tx`.
+    ///
+    /// This returns both the created task and a [TxnsCacheState] that can be
+    /// used to interact with the txn system and into which the updates should
+    /// be funneled.
+    ///
+    /// NOTE: We create both the [TxnsSubscribeTask] and the [TxnsCacheState] at
+    /// the same time because the cache is initialized with a `since_ts`, which
+    /// we get from the same [ReadHandle] that we use to initialize the
+    /// [Subscribe].
+    pub async fn open(
+        client: &PersistClient,
+        txns_id: ShardId,
+        only_data_id: Option<ShardId>,
+        tx: mpsc::Sender<TxnsReadCmd<T>>,
+    ) -> (Self, TxnsCacheState<T>) {
+        let (txns_key_schema, txns_val_schema) = C::schemas();
+        let txns_read: ReadHandle<<C as TxnsCodec>::Key, <C as TxnsCodec>::Val, T, i64> = client
+            .open_leased_reader(
+                txns_id,
+                Arc::new(txns_key_schema),
+                Arc::new(txns_val_schema),
+                Diagnostics {
+                    shard_name: "txns".to_owned(),
+                    handle_purpose: "read txns".to_owned(),
+                },
+            )
+            .await
+            .expect("txns schema shouldn't change");
+        let as_of = txns_read.since().clone();
+        let txns_id = txns_read.shard_id();
+        let since_ts = as_of.as_option().expect("txns shard is not closed").clone();
+        let txns_subscribe = txns_read
+            .subscribe(as_of)
+            .await
+            .expect("handle holds a capability");
+
+        let state = TxnsCacheState::new(txns_id, since_ts, only_data_id);
+
+        let subscribe_task = TxnsSubscribeTask {
+            txns_subscribe,
+            buf: Vec::new(),
+            tx,
+            only_data_id,
+        };
+
+        (subscribe_task, state)
+    }
+
+    async fn run(&mut self) {
+        'outer: loop {
+            let events = self.txns_subscribe.next(None).await;
+            for event in events {
+                match event {
+                    ListenEvent::Progress(frontier) => {
+                        let frontier_ts = frontier
+                            .into_option()
+                            .expect("nothing should close the txns shard");
+                        let entries = std::mem::take(&mut self.buf);
+                        let res = self
+                            .tx
+                            .send(TxnsReadCmd::Updates {
+                                entries,
+                                frontier: frontier_ts,
+                            })
+                            .await;
+                        if let Err(e) = res {
+                            debug!("exiting from TxnsSubscribeTask: {}", e);
+                            break 'outer;
```

Comment:
I usually see these sorts of loop labels as code smell. Let's move the shutting down warn up here and return instead.

Reply:
:+1: